app/controllers/users_controller.rb

```diff
@@ -858,6 +858,26 @@ def password_reset_show
     end
   end
 
+  def remove_password
+    RateLimiter.new(nil, "remove-password-hr-#{request.remote_ip}", 6, 1.hour).performed!
+    RateLimiter.new(nil, "remove-password-min-#{request.remote_ip}", 3, 1.hour).performed!
+
+    user = fetch_user_from_params
+    guardian.ensure_can_edit!(user)
+    RateLimiter.new(nil, "remove-password-hr-#{user.username}", 6, 1.hour).performed!
+
+    raise Discourse::NotFound if !user || !user.user_password
+    raise Discourse::ReadOnly if staff_writes_only_mode? && !user.staff?
+    raise Discourse::InvalidAccess if !secure_session_confirmed?
+
+    if user.associated_accounts.blank? && user.passkey_credential_ids.blank?
+      raise Discourse::InvalidAccess
+    end
+
+    user.remove_password
```

Comment:
I think maybe these validations referencing only `User` could be rolled into the `#remove_password` method. Then it's safe to use from other call sites also.